llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp

```diff
@@ -0,0 +1,726 @@
+}
+
+PreservedAnalyses
+AArch64LoopIdiomTransformPass::run(Loop &L, LoopAnalysisManager &AM,
+                                   LoopStandardAnalysisResults &AR,
+                                   LPMUpdater &) {
+  if (DisableAll)
+    return PreservedAnalyses::all();
+
+  const auto *DL = &L.getHeader()->getModule()->getDataLayout();
+
+  AArch64LoopIdiomTransform LIT(&AR.DT, &AR.LI, &AR.TTI, DL);
+  if (!LIT.run(&L))
+    return PreservedAnalyses::all();
+
+  return PreservedAnalyses::none();
+}
+
+//===----------------------------------------------------------------------===//
+//
+//          Implementation of AArch64LoopIdiomTransform
+//
+//===----------------------------------------------------------------------===//
+
+bool AArch64LoopIdiomTransform::run(Loop *L) {
+  CurLoop = L;
+
+  if (DisableAll)
+    return false;
+
+  // If the loop could not be converted to canonical form, it must have an
+  // indirectbr in it, just give up.
+  if (!L->getLoopPreheader())
+    return false;
+
+  LLVM_DEBUG(dbgs() << DEBUG_TYPE " Scanning: F["
+                    << CurLoop->getHeader()->getParent()->getName()
+                    << "] Loop %" << CurLoop->getHeader()->getName() << "\n");
+
+  return recognizeByteCompare();
+}
+
+/// Match loop-invariant value.
+template <typename SubPattern_t> struct match_LoopInvariant {
+  SubPattern_t SubPattern;
+  const Loop *L;
+
+  match_LoopInvariant(const SubPattern_t &SP, const Loop *L)
+      : SubPattern(SP), L(L) {}
+
+  template <typename ITy> bool match(ITy *V) {
+    return L->isLoopInvariant(V) && SubPattern.match(V);
+  }
+};
+
+/// Matches if the value is loop-invariant.
+template <typename Ty>
+inline match_LoopInvariant<Ty> m_LoopInvariant(const Ty &M, const Loop *L) {
+  return match_LoopInvariant<Ty>(M, L);
+}
+
+bool AArch64LoopIdiomTransform::recognizeByteCompare() {
+  if (!TTI->supportsScalableVectors() || !TTI->getMinPageSize().has_value() ||
+      DisableByteCmp)
+    return false;
+
+  BasicBlock *Header = CurLoop->getHeader();
+  BasicBlock *PH = CurLoop->getLoopPreheader();
+
+  // In AArch64LoopIdiomTransform::run we have already checked that the loop
+  // has a preheader so we can assume it's in a canonical form.
+  auto *EntryBI = cast<BranchInst>(PH->getTerminator());
+
+  if (CurLoop->getNumBackEdges() != 1 || CurLoop->getNumBlocks() != 2)
+    return false;
+
+  PHINode *PN = dyn_cast<PHINode>(&Header->front());
+  if (!PN || PN->getNumIncomingValues() != 2)
+    return false;
+
+  auto LoopBlocks = CurLoop->getBlocks();
+  // The first block in the loop should contain only 4 instructions, e.g.
+  //
+  //  while.cond:
+  //   %res.phi = phi i32 [ %start, %ph ], [ %inc, %while.body ]
+  //   %inc = add i32 %res.phi, 1
+  //   %cmp.not = icmp eq i32 %inc, %n
+  //   br i1 %cmp.not, label %while.end, label %while.body
+  //
+  auto CondBBInsts = LoopBlocks[0]->instructionsWithoutDebug();
+  if (std::distance(CondBBInsts.begin(), CondBBInsts.end()) > 4)
```

Comment:
Counting the number of instructions might be a bit too "specific". Do need to inspect the supported instructions instead?

Reply:
Is there a Godbolt link for the C leading to these instructions (I couldn't reproduce the same sequence for `compare_bytes_simple`)? I know this is to discriminate specific loop structures so chances for pattern matching later to succeed are higher, but this looks very specific to me too, and may be subject to quick changes in the future. Would dropping this lead to bottlenecks?

Reply:
Sure. https://godbolt.org/z/e5MnnGT84

As you can see, the first block corresponds to this - `while (++start != end)`, so the block is likely to be tiny. It's possible the compiler could increment `start` before the loop, in which case it will look more like `while (start++ != end)`. We've seen the former idiom in xz and 7zip. In either case the first block will have 4 or less instructions. Even if we permit more instructions we still have to match the right pattern - it would likely fail anyway because we have more than the expected uses of a value. We also couldn't, for example, permit random stores in the loop to account for the extra instructions.

I'm a bit reluctant to expand the algorithm to cover hypothetical cases that we can't prove exist. I agree it's a good idea to make it robust based on real-world examples. Do you have any examples of other specific variations of this idiom that you are worried about?

Reply:
It's worth noting that the LoopIdiomRecognition pass also uses std::distance as a way to filter out loops.

For what it's worth, I do have downstream code that increases the number of idioms we recognise. This led to me finding a number of bugs, so the exercise was very useful! However, I'd prefer to land this patch first and build upon this incrementally. I can create a patch later on that adds these extra idioms once I've done sufficient testing. I think the idioms in this patch have the biggest impact in terms of performance.

The extra idioms involve: 1) supporting 64-bit induction variable types, 2) permitting add increments without nuw and nsw flags, 3) supporting loops that increment the counter after loading the values (as opposed to currently incrementing before loading).

Reply:
Sorry for replying only now, thanks for the explanation. If I got this correctly, the snippet you provided is not meant to be transformed as the indexes of the two GEPs do not match, thus we would be bailing out. I'm confused though, why the following snippet shouldn't be supported (as per comment [here](https://github.com/llvm/llvm-project/blob/5f09a37fc4d808310b8f0ae23f10f953d05c8749/llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp#L359-L376))? Doesn't it match the specifics of the pass (the index is simply incremented before loading the values), or am I missing something?
```c
int test(unsigned char *p1, unsigned char *p2, unsigned start, unsigned end) {
  while (++start != end) {
    if (p1[start] != p2[start])
      break;
  }
  return start;
}
```
Also, if we were to support 1), 2), 3) we would still not be catching samples like https://godbolt.org/z/vPGqde4zj, due to the `and` on the induction variable (not sure why it has been converted to an i64 here, but I guess it's an orthogonal problem – provided we want to catch this).

Reply:
This logic is designed to prevent recognising loops such as `@compare_bytes_simple2`, i.e.

```
while.cond:
...
  br i1 %cmp.not, label %while.end, label %while.body

while.body:
...
  br i1 %cmp.not2, label %while.cond, label %while.end

while.end:
...
  %final_ptr = phi ptr [ %c, %while.body ], [ %d, %while.cond ]
```

where the found and end blocks are the same, and in that block there is a PHI node that has a unique incoming value for each block in the loop. However, each incoming value is not defined in the loop itself. Currently we don't handle this case in `transformByteCompare` because it would require extra work in the `byte.compare` block to select between these two incoming values. Rather than let `transformByteCompare` generate incorrect code and lead to bugs, I've explicitly disabled these cases for now. I haven't encountered any examples of such loops in the benchmarks I have looked at, so these are not critical to us. I only discovered this accidentally while writing more test cases.